New acidbombs, No iuse

pisskop opened the pull request:
Adds a general nerf to acidbombs, by adding 3 of them based on amount of liquid.  They coat anywhere from 1 tile to a random smattering to a whole 3x3 like the original.

-Micro acidbomb coats 1 tile in 1 to 2 intensity acid.  it takes just 1 flask and 250ml of acid.
-Small acidbomb uses 500ml of acid and is 2 volume.  coats a 3x3 area with 0 to 2 intesity acid.
-Large acidbombs coat a 3x3 area with 3 intensity acid like the original, but is 4 volume and 1L of acid.

As you can see, a bomb twice as large is required to achieve the same effect.  Recipe is mostly unchanged, but update to include how much acid and adds a few rags to ensure stoppers and protect against premature clinking and breaking.

I'll have to dig out the old iuse acid function, but this one removes the need for it, functions instantly, and is a bit more realistic in terms of how much acid is plausible to achieve the effect.

Coolthulhu commented on data/json/items/tools.json:
```diff
@@ -2940,38 +2940,218 @@
+        "sound_volume": 10,
+        "sound_msg": "glass shattering",
+        "no_deactivate_msg": "Its a bug that you can see this, but that doesnt mean it won't kill you.  Throw it!",
+        "fields_radius": 1.3,
```
Why that `.3`? It does nothing at all

Coolthulhu commented on data/json/items/tools.json:
```diff
@@ -2940,38 +2940,218 @@
-    "use_action": "ACIDBOMB_ACT"
+    "use_action": {
+        "type": "explosion",
+        "power": 1,
```
Why is it exploding? No need for an explosion when you just want to spread acid.

Coolthulhu commented on data/json/items/tools.json:
```diff
@@ -2940,38 +2940,218 @@
+    "weight": 1640,
+    "volume": 1,
+    "bashing": 8,
+    "to_hit": -1,
```
If this item is supposed not to appear and only be used in processing, it shouldn't have all those extra values.

Coolthulhu commented on data/json/items/tools.json:
```diff
@@ -2940,38 +2940,218 @@
-    "bashing": 4,
+    "bashing": 8,
     "to_hit": -1,
+    "flags": ["ACT_ON_RANGED_HIT", "CUSTOM_EXPLOSION"],
```
"CUSTOM_EXPLOSION" isn't used nor needed here. You'd need to define an explosion member for this item and then have it explode like a grenade. Explosion actor on the active version and ACT_ON_RANGED_HIT are enough.

Coolthulhu commented on data/json/recipes/recipe_weapon.json:
```diff
@@ -1736,30 +1736,122 @@
+    ],[
+       [ "rag", 1 ]
+    ]
+    ]
```
The spacing and newlines are hard to read, should look like:
```
"components": [
    [ [ "clay_canister", 1 ], [ "flask_glass", 1 ] ],
    [ [ "acid", 1 ] ],
    [ [ "string_6", 1 ] ],
    [ [ "rag", 1 ] ]
]
```

Coolthulhu commented on data/json/recipes/recipe_weapon.json:
```diff
@@ -1736,30 +1736,122 @@
+      [ "clay_canister", 1 ],
+      [ "flask_glass", 1 ]
+    ],[
+       [ "acid", 1 ]
```
If you're changing acid bombs, you could add some more acids here.
We have sulphuric acid (`chem_sulphuric_acid`), nitric acid (`chem_nitric_acid`) and acetic acid (`chem_acetic_acid`). They are currently quite useless, allowing them in acid bombs would give them some uses.

pisskop commented:
i was under the impression it needed to explode to paint new fields.

if it doesnt its fine

pisskop commented:
id love some new kinds of chem grenades. ill be sure to try them

Coolthulhu commented on data/json/recipes/recipe_weapon.json:
```diff
@@ -1736,30 +1736,122 @@
+      [ "bottle_glass", 1 ]
+    ],[
+      [ "clay_canister", 1 ],
+      [ "flask_glass", 1 ]
```
An extra recipe just so that volumes match is an overkill, some volume discrepancy is OK, even if it means you can somehow cram a liter of acid into a 750 ml bottle.

Coolthulhu commented on data/json/recipes/recipe_weapon.json:
```diff
@@ -1736,30 +1736,122 @@
+    ]
+    },{
+    "type" : "recipe",
+    "result": "acidbomb_medium",
```
I'd get rid of either medium or large acidbomb. They're too similar to really matter.

pisskop commented:
ahh, I see now.

I left my trash from when I was testing them in.  custom_explosion and explosion noise and et al.


I could take out the medium acidbomb and force the glass bottle to the small size.

pisskop commented:
> Why is it exploding

It needs some type of use_action, without one it returns an error.

> If you're changing acid bombs, you could add some more acids here.
We have sulphuric acid (chem_sulphuric_acid), nitric acid (chem_nitric_acid) and acetic acid (chem_acetic_acid). They are currently quite useless, allowing them in acid bombs would give them some uses.

I will get to these, especially if its as awesome as acid bombs are.  In another PR, though.

pisskop commented:
Sure.  The medium is gone.

Coolthulhu commented on data/json/items/tools.json:
```diff
@@ -2940,38 +2940,141 @@
+    "use_action": {
+        "type": "explosion",
+        "no_deactivate_msg": "Its a bug that you can see this, but that doesnt mean it won't kill you.  Throw it!",
+        "fields_radius": 0.5,
```
Test this - it will most likely not work like you expect it to.

pisskop commented:
This PR is a fix for #16808

Coolthulhu commented on data/json/recipes/recipe_weapon.json:
```diff
@@ -1736,30 +1736,71 @@
+    ],[
+      [ "rag", 1 ]
+    ]
+    ]
```
Still wrong spacing and newlines. Look at the example I provided.

pisskop commented:
>Test this - it will most likely not work like you expect it to.

It does.

<a href="http://imgur.com/POSllFI"><img src="http://i.imgur.com/POSllFI.png" title="source: imgur.com" /></a>

Coolthulhu commented on data/json/items/tools.json:
```diff
@@ -2940,38 +2940,141 @@
     "use_action": "MOLOTOV_LIT"
   },
   {
-    "id": "acidbomb",
```
You can't just delete an item like that or saves will break. You can move it into obsolete.json or change one of the bombs' id to just "acidbomb".

An item with id "acidbomb" has to exist, even if just in obsolete.json.

Coolthulhu commented:
Then specify the radius as 0. Specifying it as 0.5 will still round down to 0, but be more confusing to anyone reading the json.

Coolthulhu commented on data/json/items/tools.json:
```diff
@@ -2940,38 +2940,141 @@
+    "color": "i_green",
+    "name": "large acidbomb",
+    "description": "There is enough frothing fluid inside this acidbomb to coat a relatively large area with fields of acid.",
+    "price": 0,
```
Why price 0? Acidbombs can be useful so they should be worth few dollars.

pisskop commented:
> "components": [
    [ [ "clay_canister", 1 ], [ "flask_glass", 1 ] ],
    [ [ "acid", 1 ] ],
    [ [ "string_6", 1 ] ],
    [ [ "rag", 1 ] ]
]

This example of yours is fine, but it does not match other recipes from the file.   An example is the sling.

> 
  "components": [
    [
      [ "leather", 1 ],
      [ "fur", 1 ]
    ],
    [
      [ "plant_fibre", 20 ],
      [ "sinew", 20 ],
      [ "thread", 20 ]
    ]

Coolthulhu commented on data/json/items/tools.json:
```diff
@@ -2940,38 +2940,141 @@
+    "volume": 1,
+    "initial_charges": 1,
+    "turns_per_charge": 1,
+    "flags": ["BOMB"],
```
BOMB flag is just for radio-modable items.

Coolthulhu commented:
The file is quite old and the old syntax is hard to read. Newer files use the more readable one.
We don't have automatic styling for jsons, otherwise it would look more like `data/json/recipes/chem/fuel.json`.

BevapDin commented on data/json/items/obsolete.json:
```diff
@@ -531,5 +531,37 @@
+    "bashing": 4,
+    "to_hit": -1,
+    "category": "weapons",
+    "use_action": "ACIDBOMB_ACT"
```
You just removed the iuse function (see item_factory.cpp), items can't use it anymore.

pisskop commented:
I see.  Dont want that

Coolthulhu commented:
Instead of readding the iuse, you can just change the obsolete acid bomb to use the iuse actors.

Rivet-the-Zombie commented on data/json/items/tools.json:
```diff
@@ -2940,38 +2940,138 @@
-    "description": "This is a fragile container filled with acid.  Throw it to spill out a pool of potent acid.",
+    "color": "i_green",
+    "name": "large acidbomb",
+    "description": "A thrown acidbomb, in the process of loosing its liquid.",
```
Loosing or losing? I suppose either could work here.

pisskop commented:
Doesnt that stop things, like that acidbomb 40mm ammo, from using acidbomb then?

Coolthulhu commented:
No, 40mm acidbombs use an ammo effect to create their explosion. They do not create items on their point of impact and iuses require items to work on.

pisskop commented:
Lemme see if I get this.

-I wanted to delete the iuse function.
-I have to let the acidbomb still blow up into infinite acid.
-item_factory contains the actual iuse.
-iuse.cpp is the code that executes when the iuse is called.
-iuse.h contains the kinds of things that can call upon the iuse in question.  i.e. the actors

-I dont quite get how I would skip the iuse and tell it to use the code.  I could see making the acidbomb dump acid like these new ones do

Coolthulhu commented:
Make the old acidbomb json similar to the ones you wrote. It doesn't need to actually use the acidbomb iuse, it can use the explosion one.

Then, when acidbomb iuse is no longer used anywhere, it can be deleted from iuse.h, iuse.cpp and item_factory.cpp.

pisskop commented:
aye.

Im getting there.  slowly

Coolthulhu commented:
Needs a rebase

pisskop commented:
wth.

I updated the conflicted file, tools.json to match the vanilla master.
Then I added just the acidbomb changes, deleting the acid bomb and acidbombact.

And now theres a conflict again?  did somebody change the acidbomb?

Coolthulhu commented:
Manually updating won't help here, git tracks history.

pisskop commented:
ggggggahah

my head was just used to chop down a tree

Coolthulhu commented:
Original acidbomb was 2 acid and bottle for 3x3 acid field. I never considered it overpowered - it is quite expensive to produce since glass is limited.

Looking at the bombs here, most of them are not work producing. Micro bomb produces weak acid in small area, small acid bomb produces even weaker acid on moderate are, big bomb costs more than it needs to.

How about those stats:
* Micro producing 1x1 field of density 3, costing 1 acid and a flask, clay canister, jar or (glass) bottle
* Medium producing 3x3 field of density 3, costing 3 acid and a jar, a bottle or 3 flasks/containers
* Big producing 5x5 field of density 3, costing 5 acid and 2 jars, 2 bottles or 5 flasks/containers

And some strings/rags, of course.

That way it wouldn't be a nerf to an item that doesn't need a nerf.

Coolthulhu commented on data/json/recipes/recipe_weapon.json:
```diff
@@ -1682,30 +1682,55 @@
+    [[ "clay_canister", 1 ], [ "flask_glass", 1 ]],
+    [[ "acid", 1 ]],
+    [[ "string_6", 1 ]],
+    [[ "rag", 1 ]]
```
Formatting: needs 4 spaces before each of the lines of components

pisskop commented:
nah. a liquor store, a swamp, a chem set, and time and you have enough acidbombs to last the game.

pisskop commented:
acidbombs are almost the sole acid attack of the player  and a large reason why most critters even have an acid armor stat.  theyre silent and all you need to do is throw one at a noisemaker at night to clear a street.

Even a local one, like a standard radio that had a few firecrackers thrown at it.

pisskop commented:
the reason the micro ones are useful even at 1-2 intensity and at 1 tile is just that: they target 1 tile with low intensity that wears off fast.

  -you directly control where it goes and its perfect for sniping.
  -its impossible to miss the zombie next to you, and it will dissipate much faster than a 'pool'.
  -its considered a thrown weapon by the game, and consequently NPCs
  -flasks are common lab materials and clay is renewable


I do love acidbombs, but for skillless silent AoE attacks with common materials (water, bottles, batteries, strings) they are strong.
      the PC can act as a basic stopper for critters, abusing their cods.
     the pc can use a (potentially sacrificial) NPC to clear out a large portion of Zs
      The PC can use acid without cooking off ammo.
      The pc can make themselves immune to acid with only clothing.


Acid is the least applied armor on monsters, and  while NPCs path out of acid now most have exactly zero or 1 acid armor themselves

illi-kun commented:
> silent AoE 

It can be reasonable change to generate some noise due to glass breaking (it should be not so loud as grenade blast but definetely not completely silent)

pisskop commented:
not sure how.

it thuds, currently.


by transforming, it spawns an active bomb on a ranged hit, so the old one doesnt shatter.  the new one spawns on-tile, and explodes immedietely..

i had an explosion noise of glass breaking set but it didnt appear to activate on a charge of zero.  on 1charge it did, but then the PC usually got a turn off before.

maybe a higher cpt consumption?



id be in favor of jerry-mandering a plastic bag to hold acid...?  itd be harder though, and maybe even require kitchen buddy tools

pisskop commented:
Line 699 of ranged.cpp
https://github.com/CleverRaven/Cataclysm-DDA/blob/a7e094604b7c338f680728115264b2eacf9a16e3/src/ranged.cpp#L669

  >    // Item will shatter upon landing, destroying the item, dealing damage, and making noise

So unless thats outdated it should still be shattering?

Coolthulhu commented:
> So unless thats outdated it should still be shattering?

It only works for items with no on-drop iuse. Items with an on-drop iuse must land safely or there will be nothing to use the iuse on.

pisskop commented:
5x5 is a ridiculous field for something that wasnt propelled, imo, even discounting the logistics of liquid/field amount.

What if volume 3 750ml acidbombs were 3x3 of 3 intensity acid and the 2volume was 1-2 intensity?

Coolthulhu commented:
> What if volume 3 750ml acidbombs were 3x3 of 3 intensity acid and the 2volume was 1-2 intensity?

2 intensity bomb would be quite weak, but acceptable, I guess.

pisskop commented:
I think it has its merits in how quickly it disolves.  spitters are arguable worse than corrosives because they can cause huge area denial and because acid can hng out for so long.  The player is the mobile one in a PvZ fight, and creating areas they cant go is a disadvantage for them.

Coolthulhu commented:
* Items should not have colors with backgrounds. No other item does that.
* It is possible to activate the bomb, which will cause it to explode in inventory. Minor and unfixable on the json side, but will need to be fixed someday
* All bombs are only useful against small, weak stuff. A brute or a predator won't die to density 2 field (even if kept there for a full duration), a hulk won't notice a density 3 field etc. Anything you'd want to use the bombs on is too resistant to it

The first one needs to be fixed.
The second one can stay for now since it can't be done with jsons alone.
For the third one I'd:
* Buff medium bomb to 3x3 of density 3, to actually take down stuff
* Change big one to 5x5 of density 2 to take down or at least weaken hordes of weak crap

pisskop commented:
Acidbombs arent a 'mass murderer'  They assist in crowd control by weakening zombies so they take less hits to die.  When you have a 1hko weapon theres no reason to use acidbombs.

If we were to implement large-AoE acidbombs they should also require an explosive or some means of dissemination.  Throwing a glass bottle of liquid just doesnt make that much of a splash.  Maybe we strap it to a pipe bomb and make those actually useful?

Coolthulhu commented:
> Acidbombs arent a 'mass murderer' They assist in crowd control by weakening zombies so they take less hits to die.

They're really weak after the nerf though. Doubling the acid cost means you have to waste a whole car battery for a single bomb that actually does something (by the time you can craft acid from water, you already have better methods of mopping up trash than acid bombs).

Having to hit crappy zombies for a bit less isn't worth the expense of having to uncraft car batteries and haul around bottles. And non-crappy zombies (except shockers) resist acid too much for it to matter.

pisskop commented:
Well, in my mind, at least, acids greatest weakness is how much is needed and the lack of force.

Acidbombhacks and water cannons with reservoirs of concentrated or weak acid would replace these primative hand grenades.

The three volume bomb is still doing what the original did.  The two volume one is simply more variable.

Im no longer doubling their cost.  Im not usre why you are bothing to disassemble car batteries when standard batteries work fine?

pisskop commented:
The medium bombs are perhaps best left inefficient to encourage people to go for the bigger bang. And they still fill the role of allowing intermediate players and survivors to raid cities and kill threats that dont smash them away.  Even hunters and predators only jump around you, so it still has its uses.

The smaller acidbombs work well for newer survivors, especially if a minimal of equipment is required to use.

Coolthulhu commented:
> The three volume bomb is still doing what the original did.

The original didn't need a nerf.

> Im not sure why you are bothing to disassemble car batteries when standard batteries work fine?

Regular batteries are usually worth more than acid bombs.

> Im not sure anyone could justify a 5x5 thrown acid.

The part where it only works against trash and requires you to stand near for the zombies to stand in it.

> Its not useless or weak.

At start it isn't, but it becomes weak by the time you can mass craft it. Dangerous zombies either resist it or are too strong for it to matter.

> The medium bombs are perhaps best left inefficient to encourage people to go for the bigger bang.

More like to encourage people to skip the acid bombs altogether, since there would be no efficient acid bombs.

> Even hunters and predators only jump around you, so it still has its uses.

Wasting a turn to throw an acid bomb when there is a predator nearby is a bad tactic. You should be spending that time aiming to get a headshot when it jumps.

pisskop commented:
Well, batteries are only useful to me insofar as I need something to weld with.  MP3 is a bonus, and I only need 100 or so batteries a day for a flashlight.  The rest of the electronics are mostly wasted, with the exception of a soldering iron and a hotplate for early game use.  And even those.

A 4 volume 5 diameter bomb seems OP, but I aquiesce.

Coolthulhu commented:
> A 4 volume 5 diameter bomb seems OP

Comparable to spitter's spit, except portable, but also costly. A spit blast is relatively meaningless, except when you can't get off it for some reason.

> What are the other chemical homemade grenades going to do? White Phosphor and etc.

First it would be better to make acid bombs more useful by allowing more acids in them.

kevingranade commented:
Please don't make acid handling worse than it is in the name of balance. A
hand held container splashing to cover even one tile with enough acid to
cause significant damage is a stretch, multiple tiles is absurd.

It might be feasible if it's a puddle of acid that is fuming and creating a
cloud of acidic gas.

It is also possible that weaponizing acid vials is simply infeasible, in
which case the vials should just be dropped.

pisskop commented:
So we can agree on 

  - 1 volume, 2-3 intensity, 250ml
  - 2 volume, 1-3 intensity, 3x3 AoE 500ml
  - 3 volume 3-3 intensity, 3x3 area, 750ml

Thats more than I initially wanted, and less than you want

pisskop commented:
closing because it has produced enough discussion to warrent re-examination of the PR

kevingranade commented:
250mL is a really small volume to *cover* an entire map square, and regardless going from one square to a 3x3 AoE should imply multiplying the volume by 9, not doubling it.

Basically don't use how broken the acidbomb is to justify keeping it broken, if it's changing it should change to something that makes sense.

My suggestions, trying to reflect what you want out of it as a weapon, but doing it in a way that makes sense:
1 vol / 250mL : Acid splash, causes direct acid damage if it directly hits a target and creates a cloud of short-lived acidic gas in the target square that spreads and dissipates over a few turns.
3 vol / 750mL: "Acidic Molotov", covers one tile in an acidic puddle and created clouds of acidic gas that dissipate over 10 turns or so.  The acid puddle also produces acid clouds over time and eventually boils away. (a few minutes?)
16 vol / 4 L / 1 gal: "acid bomb", covers a 3x3 area in an acid puddle, and spawns huge volumes of acid gas for a pretty long time.  Puddle will eventually decay to the acid puddle produced by the molotov.
